Dockerfile

```diff
@@ -54,43 +73,30 @@ COPY internal/ internal/
 # Build the binary
 ENV CGO_ENABLED=1
 ARG TARGETPLATFORM
-RUN xx-go build -o source-controller -trimpath \
-    main.go
 
-FROM build-${BASE_VARIANT} as prepare-bullseye
+# The dependencies being statically built are: libgit2, libssh2, libssl, libcrypto and libz.
+# Others (such as libc, librt, libdl and libpthread) are resolved at run-time.
+# To decrease the likelihood of such dependencies being out of sync, the base build image
+# should be aligned with the target (i.e. same debian variant).
+RUN FLAGS=$(pkg-config --static --libs --cflags libssh2 libgit2 libssl libcrypto zlib openssl) \
+	xx-go build \
+        -ldflags "-s -w -extldflags \"/usr/lib/$(xx-info triple)/libssh2.a /usr/lib/$(xx-info triple)/libssl.a /usr/lib/$(xx-info triple)/libcrypto.a /usr/lib/$(xx-info triple)/libz.a -Wl,--unresolved-symbols=ignore-in-object-files -Wl,-allow-shlib-undefined ${FLAGS} -static\"" \
+        -tags 'netgo,osusergo,static_build' \
+        -o source-controller -trimpath main.go;
 
-# Move libgit2 lib to generic and predictable location
-ARG TARGETPLATFORM
-RUN mkdir -p /libgit2/lib/ \
-    && cp -d /usr/lib/$(xx-info triple)/libgit2.so* /libgit2/lib/
-
-FROM prepare-${BASE_VARIANT} as build
-
-FROM debian:${BASE_VARIANT}-slim as controller
+# The target image must aligned with apt sources used for libgit2.
+FROM debian:bookworm-slim as controller
 
 # Link repo to the GitHub Container Registry image
 LABEL org.opencontainers.image.source="https://github.com/fluxcd/source-controller"
 
-# Configure user
-RUN groupadd controller && \
-    useradd --gid controller --shell /bin/sh --create-home controller
-
-# Copy libgit2
-COPY --from=build /libgit2/lib/ /usr/local/lib/
-RUN ldconfig
-
-# Upgrade packages and install runtime dependencies
-RUN echo "deb http://deb.debian.org/debian sid main" >> /etc/apt/sources.list \
-    && echo "deb-src http://deb.debian.org/debian sid main" >> /etc/apt/sources.list \
-    && apt update \
-    && apt install --no-install-recommends -y zlib1g/sid libssl1.1/sid libssh2-1/sid \
-    && apt install --no-install-recommends -y ca-certificates \
-    && apt clean \
-    && apt autoremove --purge -y \
-    && rm -rf /var/lib/apt/lists/*
+ARG TARGETPLATFORM
+RUN apt update && apt install -y ca-certificates
 
 # Copy over binary from build
 COPY --from=build /workspace/source-controller /usr/local/bin/
+COPY --from=libgit2-bullseye /libgit2/built-on-glibc-version /
+COPY ATTRIBUTIONS.md /
 
-USER controller
+USER 65534:65534
```

Comment:
Can we change this to `nobody:nobody`? Given that people have hardcoded the `controller` user ID in their PsP, I guess this is a breaking change right?

Reply:
Reverted that, we can pick it up as a separate PR.